src/librustc_resolve/diagnostics.rs

```diff
@@ -397,6 +397,102 @@ impl Bar {
 ```
 "##,
 
+E0412: r##"
+An undeclared type name was used. Example of erroneous codes:
+
+```
+impl Something {} // error: use of undeclared type name `Something`
+// or:
+trait Foo {
+    fn bar(N); // error: use of undeclared type name `N`
+}
+```
+
+To fix this error, please verify you didn't misspell the type name,
+you did declare it or imported it into the scope. Examples:
+
+```
+struct Something;
+
+impl Something {}
+// or:
+trait Foo {
+    type N;
+
+    fn bar(Self::N);
+}
+```
+"##,
+
+E0413: r##"
+A declaration shadows an enum variant or unit-like struct in scope.
+Example of erroneous code:
+
+```
+struct Foo;
+
+let Foo = 12i32; // error: declaration of `Foo` shadows an enum variant or
+                 //        unit-like struct in scope
+```
+
+
+To fix this error, make sure declarations do not shadow any enum variants or
+structures in the scope. Please also verify that neither name was misspelled.
+Example:
+
+```
+struct Foo;
+
+let foo = 12i32; // ok!
+```
+
+Or:
+
+```
+struct FooStruct;
+
+let Foo = 12i32; // ok!
+```
+
+The goal here is to avoid a conflict of names.
+"##,
+
+E0415: r##"
+More than one function parameter have the same name. Example of erroneous
+code:
+
+```
+fn foo(f: i32, f: i32) {} // error: identifier `f` is bound more than
+                          //        once in this parameter list
+```
+
+Please verify you didn't misspell parameters' name. Example:
+
+```
+fn foo(f: i32, g: i32) {} // ok!
+```
+"##,
+
+E0416: r##"
+An identifier is bound more than once in a pattern. Example of erroneous
+code:
+
+```
+match (1, 2) {
+    (x, x) => {} // error: identifier `x` is bound more than once in the
+                 //        same pattern
+}
+```
+
```

Comment:
@GuillaumeGomez some people when first using pattern-matching think that they can write

```rust
match (A, B, C) {
    (x, x, see) => { /* A and B are equal, do one thing */ }
    (y, z, see) => { /* A and B unequal; do another thing */ }
}
```

There are programming languages (e.g. logic PL's like Prolog) that support the above.  But Rust is not one of them.

So @arielb1 is suggesting that if someone is using the same identifier more than once, they might be trying "unify" the two parts; i.e., make the pattern predicated on those two parts being equal.

A way to encode this in Rust is to use two distinct identifiers, and then add a guard to the pattern so that the arm fires only if the two identifiers are actually bound to equal values, like so:

```rust
match (A, B, C) {
    (x, x2, see) if x == x2 => { /* A and B are equal, do one thing */ }
    (y, z, see) => { /* A and B unequal; do another thing */ }
}
```